File: mobile/src/main/java/net/mm2d/dmsexplorer/viewmodel/MusicActivityModel.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.viewmodel;

import android.app.Activity;
import android.databinding.BaseObservable;
import android.databinding.Bindable;
import android.net.Uri;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.widget.Toast;

import net.mm2d.android.util.AribUtils;
import net.mm2d.android.util.Toaster;
import net.mm2d.dmsexplorer.BR;
import net.mm2d.dmsexplorer.EventLogger;
import net.mm2d.dmsexplorer.Repository;
import net.mm2d.dmsexplorer.domain.model.MediaServerModel;
import net.mm2d.dmsexplorer.domain.model.MusicPlayerModel;
import net.mm2d.dmsexplorer.domain.model.PlaybackTargetModel;
import net.mm2d.dmsexplorer.domain.model.PlayerModel;
import net.mm2d.dmsexplorer.settings.RepeatMode;
import net.mm2d.dmsexplorer.settings.Settings;
import net.mm2d.dmsexplorer.util.Downloader;
import net.mm2d.dmsexplorer.util.ThemeUtils;
import net.mm2d.dmsexplorer.view.adapter.PropertyAdapter;
import net.mm2d.dmsexplorer.viewmodel.ControlPanelModel.OnCompletionListener;
import net.mm2d.dmsexplorer.viewmodel.ControlPanelModel.SkipControlListener;
import net.mm2d.dmsexplorer.viewmodel.helper.MuteAlertHelper;

import io.reactivex.android.schedulers.AndroidSchedulers;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public class MusicActivityModel extends BaseObservable
        implements OnCompletionListener, SkipControlListener {
    @NonNull
    public final ControlPanelParam controlPanelParam;

    @NonNull
    private String mTitle;
    private int mAccentColor;
    @NonNull
    private ControlPanelModel mControlPanelModel;
    @NonNull
    private PropertyAdapter mPropertyAdapter;
    @Nullable
    private byte[] mImageBinary;

    @NonNull
    private RepeatMode mRepeatMode;
    @DrawableRes
    private int mRepeatIconId;
    @Nullable
    private Toast mToast;

    @NonNull
    private final Activity mActivity;
    @NonNull
    private final Repository mRepository;
    @NonNull
    private final MediaServerModel mServerModel;
    @NonNull
    private final Settings mSettings;
    @NonNull
    private final MuteAlertHelper mMuteAlertHelper;

    public MusicActivityModel(
            @NonNull final Activity activity,
            @NonNull final Repository repository) {
        mActivity = activity;
        mRepository = repository;
        mServerModel = repository.getMediaServerModel();
        mSettings = new Settings(activity);
        mRepeatMode = mSettings.getRepeatModeMusic();
        mRepeatIconId = mRepeatMode.getIconId();

        controlPanelParam = new ControlPanelParam();
        mMuteAlertHelper = new MuteAlertHelper(activity);
        updateTargetModel();
    }

    private void updateTargetModel() {
        final PlaybackTargetModel targetModel = mRepository.getPlaybackTargetModel();
        if (targetModel == null || targetModel.getUri() == Uri.EMPTY) {
            ActivityCompat.finishAfterTransition(mActivity);
            return;
        }
        mMuteAlertHelper.alertIfMuted();
        final PlayerModel playerModel = new MusicPlayerModel(mActivity);
        mControlPanelModel = new ControlPanelModel(mActivity, playerModel);
        mControlPanelModel.setRepeatMode(mRepeatMode);
        mControlPanelModel.setOnCompletionListener(this);
        mControlPanelModel.setSkipControlListener(this);
        playerModel.setUri(targetModel.getUri(), null);

        mTitle = AribUtils.toDisplayableString(targetModel.getTitle());
        mAccentColor = ThemeUtils.getDeepColor(mTitle);
        mPropertyAdapter = PropertyAdapter.ofContent(mActivity, targetModel.getContentEntity());
        mRepository.getThemeModel().setThemeColor(mActivity, mAccentColor, 0);

        notifyPropertyChanged(BR.title);
        notifyPropertyChanged(BR.accentColor);
        notifyPropertyChanged(BR.propertyAdapter);
        notifyPropertyChanged(BR.controlPanelModel);

        controlPanelParam.setBackgroundColor(mAccentColor);

        loadArt(targetModel.getContentEntity().getArtUri());
    }

    private void loadArt(@NonNull final Uri uri) {
        setImageBinary(null);
        if (uri == Uri.EMPTY) {
            return;
        }
        Downloader.create(uri.toString())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setImageBinary);
    }

    public void terminate() {
        mControlPanelModel.terminate();
    }

    public void restoreSaveProgress(final int position) {
        mControlPanelModel.restoreSaveProgress(position);
    }

    public int getCurrentProgress() {
        return mControlPanelModel.getProgress();
    }

    public void onClickRepeat() {
        mRepeatMode = mRepeatMode.next();
        mControlPanelModel.setRepeatMode(mRepeatMode);
        setRepeatIconId(mRepeatMode.getIconId());
        mSettings.setRepeatModeMusic(mRepeatMode);

        showRepeatToast();
    }

    private void showRepeatToast() {
        if (mToast != null) {
            mToast.cancel();
        }
        mToast = Toaster.showLong(mActivity, mRepeatMode.getMessageId());
    }

    @Bindable
    public int getRepeatIconId() {
        return mRepeatIconId;
    }

    public void setRepeatIconId(@DrawableRes final int id) {
        mRepeatIconId = id;
        notifyPropertyChanged(BR.repeatIconId);
    }

    @Nullable
    @Bindable
    public byte[] getImageBinary() {
        return mImageBinary;
    }

    public void setImageBinary(@Nullable final byte[] imageBinary) {
        mImageBinary = imageBinary;
        notifyPropertyChanged(BR.imageBinary);
    }

    @NonNull
    @Bindable
    public String getTitle() {
        return mTitle;
    }

    @Bindable
    public int getAccentColor() {
        return mAccentColor;
    }

    @NonNull
    @Bindable
    public PropertyAdapter getPropertyAdapter() {
        return mPropertyAdapter;
    }

    @NonNull
    @Bindable
    public ControlPanelModel getControlPanelModel() {
        return mControlPanelModel;
    }

    @Override
    public void onCompletion() {
        mControlPanelModel.terminate();
        if (!selectNext()) {
            ActivityCompat.finishAfterTransition(mActivity);
            return;
        }
        updateTargetModel();
        EventLogger.sendPlayContent(true);
    }

    @Override
    public void next() {
        mControlPanelModel.terminate();
        if (!selectNext()) {
            ActivityCompat.finishAfterTransition(mActivity);
            return;
        }
        updateTargetModel();
        EventLogger.sendPlayContent(true);
    }

    @Override
    public void previous() {
        mControlPanelModel.terminate();
        if (!selectPrevious()) {
            ActivityCompat.finishAfterTransition(mActivity);
            return;
        }
        updateTargetModel();
        EventLogger.sendPlayContent(true);
    }

    private boolean selectNext() {
        switch (mRepeatMode) {
            case PLAY_ONCE:
                return false;
            case SEQUENTIAL:
                return mServerModel.selectNextEntity(MediaServerModel.SCAN_MODE_SEQUENTIAL);
            case REPEAT_ALL:
                return mServerModel.selectNextEntity(MediaServerModel.SCAN_MODE_LOOP);
            case REPEAT_ONE:
                return false;
        }
        return false;
    }

    private boolean selectPrevious() {
        switch (mRepeatMode) {
            case PLAY_ONCE:
                return false;
            case SEQUENTIAL:
                return mServerModel.selectPreviousEntity(MediaServerModel.SCAN_MODE_SEQUENTIAL);
            case REPEAT_ALL:
                return mServerModel.selectPreviousEntity(MediaServerModel.SCAN_MODE_LOOP);
            case REPEAT_ONE:
                return false;
        }
        return false;
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/viewmodel/MusicActivityModel.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.viewmodel

import android.net.Uri
import android.widget.Toast
import androidx.core.app.ActivityCompat
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import net.mm2d.android.util.AribUtils
import net.mm2d.android.util.Toaster
import net.mm2d.dmsexplorer.Repository
import net.mm2d.dmsexplorer.domain.model.MediaServerModel
import net.mm2d.dmsexplorer.domain.model.MusicPlayerModel
import net.mm2d.dmsexplorer.log.EventLogger
import net.mm2d.dmsexplorer.settings.RepeatMode
import net.mm2d.dmsexplorer.settings.Settings
import net.mm2d.dmsexplorer.util.Downloader
import net.mm2d.dmsexplorer.view.adapter.PropertyAdapter
import net.mm2d.dmsexplorer.view.base.BaseActivity
import net.mm2d.dmsexplorer.viewmodel.helper.MuteAlertHelper

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class MusicActivityModel(
    private val activity: BaseActivity,
    private val repository: Repository,
) {
    private val serverModel: MediaServerModel? = repository.mediaServerModel
    private val settings: Settings = Settings.get()
    private var repeatMode: RepeatMode = settings.repeatModeMusic
    private var toast: Toast? = null
    private val muteAlertHelper: MuteAlertHelper = MuteAlertHelper(activity)
    private var playStartTime: Long = 0
    private var finishing: Boolean = false
    val controlPanelParam: ControlPanelParam = ControlPanelParam()

    private val titleFlow: MutableStateFlow<String> = MutableStateFlow("")
    fun getTitleFlow(): Flow<String> = titleFlow

    private val controlColorFlow: MutableStateFlow<Int> = MutableStateFlow(0)
    fun getControlColorFlow(): Flow<Int> = controlColorFlow

    private val controlPanelModelFlow: MutableStateFlow<ControlPanelModel> =
        MutableStateFlow(ControlPanelModel(activity, MusicPlayerModel(activity)))

    fun getControlPanelModelFlow(): Flow<ControlPanelModel> = controlPanelModelFlow
    private var controlPanelModel: ControlPanelModel
        get() = controlPanelModelFlow.value
        private set(value) {
            controlPanelModelFlow.value = value
        }

    private val propertyAdapterFlow: MutableSharedFlow<PropertyAdapter> =
        MutableSharedFlow(replay = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)

    fun getPropertyAdapterFlow(): Flow<PropertyAdapter> = propertyAdapterFlow
    private var propertyAdapter: PropertyAdapter
        get() = propertyAdapterFlow.replayCache.first()
        set(value) {
            propertyAdapterFlow.tryEmit(value)
        }

    private val imageBinaryFlow: MutableStateFlow<ByteArray?> = MutableStateFlow(null)
    fun getImageBinaryFlow(): Flow<ByteArray?> = imageBinaryFlow

    val currentProgress: Int
        get() = controlPanelModel.getProgress()

    private val repeatIconIdFlow: MutableStateFlow<Int> = MutableStateFlow(repeatMode.iconId)
    fun getRepeatIconIdFlow(): Flow<Int> = repeatIconIdFlow

    private val isTooShortPlayTime: Boolean
        get() {
            val playTime = System.currentTimeMillis() - playStartTime
            return !controlPanelModel.isSkipped && playTime < TOO_SHORT_PLAY_TIME
        }

    init {
        val targetModel = repository.playbackTargetModel
        check(!(targetModel == null || targetModel.uri === Uri.EMPTY))
        updateTargetModel()
    }

    private fun updateTargetModel() {
        val targetModel = repository.playbackTargetModel
        if (targetModel == null || targetModel.uri === Uri.EMPTY) {
            finishAfterTransition()
            return
        }
        playStartTime = System.currentTimeMillis()
        muteAlertHelper.alertIfMuted()
        val playerModel = MusicPlayerModel(activity)
        controlPanelModel = ControlPanelModel(activity, playerModel).also {
            it.setRepeatMode(repeatMode)
            it.setOnCompletionListener(this::onCompletion)
            it.setSkipControlListener(this::onNext, this::onPrevious)
        }
        playerModel.setUri(targetModel.uri, null)

        titleFlow.value = AribUtils.toDisplayableString(targetModel.title)
        val generator = settings
            .themeParams
            .themeColorGenerator
        controlColorFlow.value = generator.getControlColor(titleFlow.value)
        propertyAdapter = PropertyAdapter.ofContent(activity, targetModel.contentEntity)
        repository.themeModel.setThemeColor(activity, controlColorFlow.value, 0)

        controlPanelParam.setBackgroundColor(controlColorFlow.value)

        loadArt(targetModel.contentEntity.artUri)
    }

    private fun loadArt(
        uri: Uri,
    ) {
        imageBinaryFlow.value = null
        if (uri === Uri.EMPTY) {
            return
        }
        Downloader.create(uri.toString())
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                Consumer { imageBinaryFlow.value = it },
            ).isDisposed
    }

    fun terminate() {
        controlPanelModel.terminate()
    }

    fun restoreSaveProgress(
        position: Int,
    ) {
        controlPanelModel.restoreSaveProgress(position)
    }

    fun onClickRepeat() {
        repeatMode = repeatMode.next()
        controlPanelModel.setRepeatMode(repeatMode)
        repeatIconIdFlow.value = repeatMode.iconId
        settings.repeatModeMusic = repeatMode

        showRepeatToast()
    }

    private fun showRepeatToast() {
        toast?.cancel()
        toast = Toaster.show(activity, repeatMode.messageId)
    }

    private fun onCompletion() {
        controlPanelModel.terminate()
        if (isTooShortPlayTime || controlPanelModel.hasError() || !selectNext()) {
            finishAfterTransition()
            return
        }
        updateTargetModel()
        EventLogger.sendPlayContent(true)
    }

    private fun onNext() {
        controlPanelModel.terminate()
        if (!selectNext()) {
            finishAfterTransition()
            return
        }
        updateTargetModel()
        EventLogger.sendPlayContent(true)
    }

    private fun onPrevious() {
        controlPanelModel.terminate()
        if (!selectPrevious()) {
            finishAfterTransition()
            return
        }
        updateTargetModel()
        EventLogger.sendPlayContent(true)
    }

    private fun finishAfterTransition() {
        if (!finishing) {
            finishing = true
            ActivityCompat.finishAfterTransition(activity)
        }
    }

    private fun selectNext(): Boolean =
        when (repeatMode) {
            RepeatMode.PLAY_ONCE -> false
            RepeatMode.SEQUENTIAL ->
                serverModel?.selectNextEntity(MediaServerModel.SCAN_MODE_SEQUENTIAL) ?: false

            RepeatMode.REPEAT_ALL ->
                serverModel?.selectNextEntity(MediaServerModel.SCAN_MODE_LOOP) ?: false

            RepeatMode.REPEAT_ONE -> false
        }

    private fun selectPrevious(): Boolean =
        when (repeatMode) {
            RepeatMode.PLAY_ONCE -> false
            RepeatMode.SEQUENTIAL ->
                serverModel?.selectPreviousEntity(MediaServerModel.SCAN_MODE_SEQUENTIAL) ?: false

            RepeatMode.REPEAT_ALL ->
                serverModel?.selectPreviousEntity(MediaServerModel.SCAN_MODE_LOOP) ?: false

            RepeatMode.REPEAT_ONE -> false
        }

    companion object {
        private const val TOO_SHORT_PLAY_TIME = 2000L
    }
}
